src/GaIA/pkgs/perl/ActivePerl-5.8.7.815-i686-linux-2.2.17-gcc-211909/perl/man/man3/Config.3
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.3
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "Config 3"
.TH Config 3 "2005-06-22" "perl v5.8.7" "Perl Programmers Reference Guide"
.SH "NAME"
Config \- access Perl configuration information
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 4
\&    use Config;
\&    if ($Config{usethreads}) {
\&        print "has thread support\en"
\&    }
.Ve
.PP
.Vb 1
\&    use Config qw(myconfig config_sh config_vars config_re);
.Ve
.PP
.Vb 1
\&    print myconfig();
.Ve
.PP
.Vb 1
\&    print config_sh();
.Ve
.PP
.Vb 1
\&    print config_re();
.Ve
.PP
.Vb 1
\&    config_vars(qw(osname archname));
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The Config module contains all the information that was available to
the \f(CW\*(C`Configure\*(C'\fR program at Perl build time (over 900 values).
.PP
Shell variables from the \fIconfig.sh\fR file (written by Configure) are
stored in the readonly-variable \f(CW%Config\fR, indexed by their names.
.PP
Values stored in config.sh as 'undef' are returned as undefined
values.  The perl \f(CW\*(C`exists\*(C'\fR function can be used to check if a
named variable exists.
.IP "\fImyconfig()\fR" 4
.IX Item "myconfig()"
Returns a textual summary of the major perl configuration values.
See also \f(CW\*(C`\-V\*(C'\fR in \*(L"Switches\*(R" in perlrun.
.IP "\fIconfig_sh()\fR" 4
.IX Item "config_sh()"
Returns the entire perl configuration information in the form of the
original config.sh shell variable assignment script.
.IP "config_re($regex)" 4
.IX Item "config_re($regex)"
Like \fIconfig_sh()\fR but returns, as a list, only the config entries who's
names match the \f(CW$regex\fR.
.IP "config_vars(@names)" 4
.IX Item "config_vars(@names)"
Prints to \s-1STDOUT\s0 the values of the named configuration variable. Each is
printed on a separate line in the form:
.Sp
.Vb 1
\&  name='value';
.Ve
.Sp
Names which are unknown are output as \f(CW\*(C`name='UNKNOWN';\*(C'\fR.
See also \f(CW\*(C`\-V:name\*(C'\fR in \*(L"Switches\*(R" in perlrun.
.SH "EXAMPLE"
.IX Header "EXAMPLE"
Here's a more sophisticated example of using \f(CW%Config:\fR
.PP
.Vb 2
\&    use Config;
\&    use strict;
.Ve
.PP
.Vb 11
\&    my %sig_num;
\&    my @sig_name;
\&    unless($Config{sig_name} && $Config{sig_num}) {
\&        die "No sigs?";
\&    } else {
\&        my @names = split ' ', $Config{sig_name};
\&        @sig_num{@names} = split ' ', $Config{sig_num};
\&        foreach (@names) {
\&            $sig_name[$sig_num{$_}] ||= $_;
\&        }   
\&    }
.Ve
.PP
.Vb 4
\&    print "signal #17 = $sig_name[17]\en";
\&    if ($sig_num{ALRM}) { 
\&        print "SIGALRM is $sig_num{ALRM}\en";
\&    }
.Ve
.SH "WARNING"
.IX Header "WARNING"
Because this information is not stored within the perl executable
itself it is possible (but unlikely) that the information does not
relate to the actual perl binary which is being used to access it.
.PP
The Config module is installed into the architecture and version
specific library directory ($Config{installarchlib}) and it checks the
perl version number when loaded.
.PP
The values stored in config.sh may be either single-quoted or
double\-quoted. Double-quoted strings are handy for those cases where you
need to include escape sequences in the strings. To avoid runtime variable
interpolation, any \f(CW\*(C`$\*(C'\fR and \f(CW\*(C`@\*(C'\fR characters are replaced by \f(CW\*(C`\e$\*(C'\fR and
\&\f(CW\*(C`\e@\*(C'\fR, respectively. This isn't foolproof, of course, so don't embed \f(CW\*(C`\e$\*(C'\fR
or \f(CW\*(C`\e@\*(C'\fR in double-quoted strings unless you're willing to deal with the
consequences. (The slashes will end up escaped and the \f(CW\*(C`$\*(C'\fR or \f(CW\*(C`@\*(C'\fR will
trigger variable interpolation)
.SH "GLOSSARY"
.IX Header "GLOSSARY"
Most \f(CW\*(C`Config\*(C'\fR variables are determined by the \f(CW\*(C`Configure\*(C'\fR script
on platforms supported by it (which is most \s-1UNIX\s0 platforms).  Some
platforms have custom-made \f(CW\*(C`Config\*(C'\fR variables, and may thus not have
some of the variables described below, or may have extraneous variables
specific to that particular port.  See the port specific documentation
in such cases.
.Sh "_"
.IX Subsection "_"
.ie n .IP """_a""" 4
.el .IP "\f(CW_a\fR" 4
.IX Item "_a"
From \fIUnix.U\fR:
.Sp
This variable defines the extension used for ordinary library files.
For unix, it is \fI.a\fR.  The \fI.\fR is included.  Other possible
values include \fI.lib\fR.
.ie n .IP """_exe""" 4
.el .IP "\f(CW_exe\fR" 4
.IX Item "_exe"
From \fIUnix.U\fR:
.Sp
This variable defines the extension used for executable files.
\&\f(CW\*(C`DJGPP\*(C'\fR, Cygwin and \fI\s-1OS/2\s0\fR use \fI.exe\fR.  Stratus \f(CW\*(C`VOS\*(C'\fR uses \fI.pm\fR.
On operating systems which do not require a specific extension
for executable files, this variable is empty.
.ie n .IP """_o""" 4
.el .IP "\f(CW_o\fR" 4
.IX Item "_o"
From \fIUnix.U\fR:
.Sp
This variable defines the extension used for object files.
For unix, it is \fI.o\fR.  The \fI.\fR is included.  Other possible
values include \fI.obj\fR.
.Sh "a"
.IX Subsection "a"
.ie n .IP """afs""" 4
.el .IP "\f(CWafs\fR" 4
.IX Item "afs"
From \fIafs.U\fR:
.Sp
This variable is set to \f(CW\*(C`true\*(C'\fR if \f(CW\*(C`AFS\*(C'\fR (Andrew File System) is used
on the system, \f(CW\*(C`false\*(C'\fR otherwise.  It is possible to override this
with a hint value or command line option, but you'd better know
what you are doing.
.ie n .IP """afsroot""" 4
.el .IP "\f(CWafsroot\fR" 4
.IX Item "afsroot"
From \fIafs.U\fR:
.Sp
This variable is by default set to \fI/afs\fR. In the unlikely case
this is not the correct root, it is possible to override this with
a hint value or command line option.  This will be used in subsequent
tests for AFSness in the Perl configure and test process. 
.ie n .IP """alignbytes""" 4
.el .IP "\f(CWalignbytes\fR" 4
.IX Item "alignbytes"
From \fIalignbytes.U\fR:
.Sp
This variable holds the number of bytes required to align a
double\*(-- or a long double when applicable. Usual values are
2, 4 and 8.  The default is eight, for safety.
.ie n .IP """ansi2knr""" 4
.el .IP "\f(CWansi2knr\fR" 4
.IX Item "ansi2knr"
From \fIansi2knr.U\fR:
.Sp
This variable is set if the user needs to run ansi2knr.  
Currently, this is not supported, so we just abort.
.ie n .IP """aphostname""" 4
.el .IP "\f(CWaphostname\fR" 4
.IX Item "aphostname"
From \fId_gethname.U\fR:
.Sp
This variable contains the command which can be used to compute the
host name. The command is fully qualified by its absolute path, to make
it safe when used by a process with super-user privileges.
.ie n .IP """api_revision""" 4
.el .IP "\f(CWapi_revision\fR" 4
.IX Item "api_revision"
From \fIpatchlevel.U\fR:
.Sp
The three variables, api_revision, api_version, and
api_subversion, specify the version of the oldest perl binary
compatible with the present perl.  In a full version string
such as \fI5.6.1\fR, api_revision is the \f(CW5\fR.
Prior to 5.5.640, the format was a floating point number,
like 5.00563.
.Sp
\&\fIperl.c\fR:\fIincpush()\fR and \fIlib/lib.pm\fR will automatically search in
\&\fI$sitelib/.\fR. for older directories back to the limit specified
by these api_ variables.  This is only useful if you have a
perl library directory tree structured like the default one.
See \f(CW\*(C`INSTALL\*(C'\fR for how this works.  The versioned site_perl
directory was introduced in 5.005, so that is the lowest
possible value.  The version list appropriate for the current
system is determined in \fIinc_version_list.U\fR.
.Sp
\&\f(CW\*(C`XXX\*(C'\fR To do:  Since compatibility can depend on compile time
options (such as bincompat, longlong, etc.) it should
(perhaps) be set by Configure, but currently it isn't.
Currently, we read a hard-wired value from \fIpatchlevel.h\fR.
Perhaps what we ought to do is take the hard-wired value from
\&\fIpatchlevel.h\fR but then modify it if the current Configure
options warrant.  \fIpatchlevel.h\fR then would use an #ifdef guard.
.ie n .IP """api_subversion""" 4
.el .IP "\f(CWapi_subversion\fR" 4
.IX Item "api_subversion"
From \fIpatchlevel.U\fR:
.Sp
The three variables, api_revision, api_version, and
api_subversion, specify the version of the oldest perl binary
compatible with the present perl.  In a full version string
such as \fI5.6.1\fR, api_subversion is the \f(CW1\fR.  See api_revision for
full details.
.ie n .IP """api_version""" 4
.el .IP "\f(CWapi_version\fR" 4
.IX Item "api_version"
From \fIpatchlevel.U\fR:
.Sp
The three variables, api_revision, api_version, and
api_subversion, specify the version of the oldest perl binary
compatible with the present perl.  In a full version string
such as \fI5.6.1\fR, api_version is the \f(CW6\fR.  See api_revision for
full details.  As a special case, 5.5.0 is rendered in the
old-style as 5.005.  (In the 5.005_0x maintenance series,
this was the only versioned directory in \f(CW$sitelib\fR.)
.ie n .IP """api_versionstring""" 4
.el .IP "\f(CWapi_versionstring\fR" 4
.IX Item "api_versionstring"
From \fIpatchlevel.U\fR:
.Sp
This variable combines api_revision, api_version, and
api_subversion in a format such as 5.6.1 (or 5_6_1) suitable
for use as a directory name.  This is filesystem dependent.
.ie n .IP """ar""" 4
.el .IP "\f(CWar\fR" 4
.IX Item "ar"
From \fILoc.U\fR:
.Sp
This variable is used internally by Configure to determine the
full pathname (if any) of the ar program.  After Configure runs,
the value is reset to a plain \f(CW\*(C`ar\*(C'\fR and is not useful.
.ie n .IP """archlib""" 4
.el .IP "\f(CWarchlib\fR" 4
.IX Item "archlib"
From \fIarchlib.U\fR:
.Sp
This variable holds the name of the directory in which the user wants
to put architecture-dependent public library files for \f(CW$package\fR.
It is most often a local directory such as \fI/usr/local/lib\fR.
Programs using this variable must be prepared to deal
with filename expansion.
.ie n .IP """archlibexp""" 4
.el .IP "\f(CWarchlibexp\fR" 4
.IX Item "archlibexp"
From \fIarchlib.U\fR:
.Sp
This variable is the same as the archlib variable, but is
filename expanded at configuration time, for convenient use.
.ie n .IP """archname64""" 4
.el .IP "\f(CWarchname64\fR" 4
.IX Item "archname64"
From \fIuse64bits.U\fR:
.Sp
This variable is used for the 64\-bitness part of \f(CW$archname\fR.
.ie n .IP """archname""" 4
.el .IP "\f(CWarchname\fR" 4
.IX Item "archname"
From \fIarchname.U\fR:
.Sp
This variable is a short name to characterize the current
architecture.  It is used mainly to construct the default archlib.
.ie n .IP """archobjs""" 4
.el .IP "\f(CWarchobjs\fR" 4
.IX Item "archobjs"
From \fIUnix.U\fR:
.Sp
This variable defines any additional objects that must be linked
in with the program on this architecture.  On unix, it is usually
empty.  It is typically used to include emulations of unix calls
or other facilities.  For perl on \fI\s-1OS/2\s0\fR, for example, this would
include \fIos2/os2.obj\fR.
.ie n .IP """asctime_r_proto""" 4
.el .IP "\f(CWasctime_r_proto\fR" 4
.IX Item "asctime_r_proto"
From \fId_asctime_r.U\fR:
.Sp
This variable encodes the prototype of asctime_r.
It is zero if d_asctime_r is undef, and one of the
\&\f(CW\*(C`REENTRANT_PROTO_T_ABC\*(C'\fR macros of \fIreentr.h\fR if d_asctime_r
is defined.
.ie n .IP """awk""" 4
.el .IP "\f(CWawk\fR" 4
.IX Item "awk"
From \fILoc.U\fR:
.Sp
This variable is used internally by Configure to determine the
full pathname (if any) of the awk program.  After Configure runs,
the value is reset to a plain \f(CW\*(C`awk\*(C'\fR and is not useful.
.Sh "b"
.IX Subsection "b"
.ie n .IP """baserev""" 4
.el .IP "\f(CWbaserev\fR" 4
.IX Item "baserev"
From \fIbaserev.U\fR:
.Sp
The base revision level of this package, from the \fI.package\fR file.
.ie n .IP """bash""" 4
.el .IP "\f(CWbash\fR" 4
.IX Item "bash"
From \fILoc.U\fR:
.Sp
This variable is defined but not used by Configure.
The value is a plain '' and is not useful.
.ie n .IP """bin""" 4
.el .IP "\f(CWbin\fR" 4
.IX Item "bin"
From \fIbin.U\fR:
.Sp
This variable holds the name of the directory in which the user wants
to put publicly executable images for the package in question.  It
is most often a local directory such as \fI/usr/local/bin\fR. Programs using
this variable must be prepared to deal with \fI~name\fR substitution.
.ie n .IP """binexp""" 4
.el .IP "\f(CWbinexp\fR" 4
.IX Item "binexp"
From \fIbin.U\fR:
.Sp
This is the same as the bin variable, but is filename expanded at
configuration time, for use in your makefiles.
.ie n .IP """bison""" 4
.el .IP "\f(CWbison\fR" 4
.IX Item "bison"
From \fILoc.U\fR:
.Sp
This variable is used internally by Configure to determine the
full pathname (if any) of the bison program.  After Configure runs,
the value is reset to a plain \f(CW\*(C`bison\*(C'\fR and is not useful.
.ie n .IP """byacc""" 4
.el .IP "\f(CWbyacc\fR" 4
.IX Item "byacc"
From \fILoc.U\fR:
.Sp
This variable is used internally by Configure to determine the
full pathname (if any) of the byacc program.  After Configure runs,
the value is reset to a plain \f(CW\*(C`byacc\*(C'\fR and is not useful.
.ie n .IP """byteorder""" 4
.el .IP "\f(CWbyteorder\fR" 4
.IX Item "byteorder"
From \fIbyteorder.U\fR:
.Sp
This variable holds the byte order in a \f(CW\*(C`UV\*(C'\fR. In the following,
larger digits indicate more significance.  The variable byteorder

src/GaIA/UNItools_buildroot/var/opt/UNItools/lib/site_perl/5.8.7/PPM/Config.pm
package PPM::Config;

use strict;
use Config;
use Data::Dumper;
use File::Path;
require PPM::YAML;

$PPM::Config::VERSION = '3.00';

use constant AIX         => 'aix'     eq $^O;
use constant SOLARIS     => 'solaris' eq $^O;
use constant SOLARIS_X86 => SOLARIS   && $Config{archname} =~ /i86pc|x86/;
use constant WIN32       => 'MSWin32' eq $^O;

sub new {
    my $class = shift;
    my $self = bless { }, ref($class) || $class;
    my $file = shift;
    $self->{DATA} = {};
    if (defined $file) {
	$self->loadfile($file, 'load');
	$self->setfile($file);
	$self->setsave;
    }
    return $self;
}

sub config {
    my $o = shift;
    return wantarray ? %{$o->{DATA}} : $o->{DATA};
}

sub loadfile {
    my $o = shift;
    my $file = shift;
    my $action = shift;
    print "DEBUG: Loading file: $file.\n" if $ENV{PPM3_CONFIG_DEBUG};
    open(FILE, "< $file")		|| die "can't read $file: $!";
    my $str = do { local $/; <FILE> };
    my $dat = eval { PPM::YAML::deserialize($str) } || {};
    close(FILE)				|| die "can't close $file: $!";
    $o->load($dat, $action);
    $o;
}

sub load {
    my $o = shift;
    my $dat = shift;
    my $action = shift || 'load';
    if ($action eq 'load' or not exists $o->{DATA}) {
	$o->{DATA} = $dat;
    }
    else {
	$o->merge($dat);
    }
    $o;
}

sub file { $_[0]->{file} }

sub setfile {
    my $o = shift;
    my $file = shift;
    $o->{file} = $file;
    $o;
}

sub setsave {
    my $o = shift;
    $o->{autosave} = 1;
    $o;
}

sub save {
    my $o = shift;
    my $file = shift || $o->{file};
    my $mode = (stat($file))[2] & 07777;
    $mode |= 0222;      # turn on write permissions (if not already)
    chmod $mode, $file; # ignore failures
    open(FILE, "> $file") or do {
	print STDERR <<END;
Warning: save $file: $!.
    => Configuration not saved.
END
	return;
    };
    my $str = PPM::YAML::serialize($o->{DATA});
    print FILE $str;
    close(FILE)				|| die "can't close $file: $!";
    $o;
}

sub merge {
    my $o = shift;
    my $dat = shift;
    _merge(\$o->{DATA}, \$dat)
      if (defined $dat);
    $o;
}

sub DESTROY {
    my $o = shift;
    $o->save if $o->{autosave};
}

sub _merge {
    my ($old_ref, $new_ref) = @_;

    return unless defined $old_ref and defined $new_ref;

    my $r_old = ref($old_ref);
    my $r_new = ref($new_ref);

    return unless $r_old eq $r_new;
    
    if ($r_old eq 'SCALAR') {
	$$old_ref = $$new_ref;
    }
    elsif ($r_old eq 'REF') {
	my $old = $$old_ref;
	my $new = $$new_ref;
	$r_old = ref($old);
	$r_new = ref($new);

	return unless $r_old eq $r_new;

	if (ref($old) eq 'HASH') {
	    for my $key (keys %$new) {
		if (exists $old->{$key} and
		    defined $old->{$key} and
		    defined $new->{$key}) {
		    _merge(\$old->{$key}, \$new->{$key});
		}
		else {
		    $old->{$key} = $new->{$key};
		}
	    }
	}
	elsif (ref($old) eq 'ARRAY') {
	    for my $item (@$new) {
		if (ref($item) eq '' and not grep { $item eq $_ } @$old) {
		    push @$old, $item;
		}
		elsif(ref($item)) {
		    push @$old, $item;
		}
	    }
	}
    }
}

#=============================================================================
# get_conf_dirs(): return a list of directories to search for config files.
#=============================================================================
use constant DELIM	=> WIN32 ? ';' : ':';
use constant PATHSEP	=> WIN32 ? '\\' : '/';
use constant KEYDIR	=> 'ActiveState';
use constant KEYFILE	=> 'ActiveState.lic';
use constant CONFDIR	=> 'PPM';
use constant CONFIG_SUFFIX => '.cfg';
use constant UNIX_SHARED_ROOT => '/usr/local/etc';

sub mymkpath {
    my $path = shift;
    unless (-d $path) {
	mkpath($path);
	die "Couldn't create directory $path: $!"
	  unless -d $path;
    }
    $path;
}

sub get_license_file {
    my $license_dir = licGetHomeDir();
    my $lic_file = join PATHSEP, $license_dir, KEYFILE;
    return $lic_file;
}

BEGIN {
    if ($ENV{PPM3_CONFIG_DEBUG}) {
	my $shared = $ENV{PPM3_SHARED} ? '' : 'not ';
	my $user   = $ENV{PPM3_USER}   ? '' : 'not ';
	print <<END;
DEBUG: Will ${shared}stat shared configuration files...
DEBUG: Will ${user}stat user's configuration files...
END
    }
}

sub load_config_file {
    my $orig = shift;
    my $mode = shift || 'rw'; # 'ro' for read-only.

    my $name = $orig . CONFIG_SUFFIX;
    my $conf = PPM::Config->new;

    # Load all config files in the "configuration path"
    my $treedir = eval { get_tree_conf_dir()   };
    my $userdir = eval { get_user_conf_dir()   };
    my $shrddir = eval { get_shared_conf_dir() };
    unless (grep { defined $_ } ($userdir, $shrddir, $treedir)) {
	print <<END;

    *** FATAL ERROR *** 
    
    Couldn't find the PPM configuration directories in either your home
    directory or the shared directory. That probably means neither of the
    environment variables PPM3_SHARED and PPM3_USER were set by the wrapper
    program "ppm3".

    Set the environment variable PPM3_CONFIG_DEBUG to 1, then rerun PPM
    to get more diagnostics about where it loaded its initial
    configuration.

END
	exit(1);
    }
    my ($treefile, $userfile, $shrdfile);
    $treefile = defined $treedir ? join PATHSEP, $treedir, $name : '';
    $userfile = defined $userdir ? join PATHSEP, $userdir, $name : '';
    $shrdfile = defined $shrddir ? join PATHSEP, $shrddir, $name : '';
    print "DEBUG: treefile='$treefile'\n" if $ENV{PPM3_CONFIG_DEBUG};
    print "DEBUG: userfile='$userfile'\n" if $ENV{PPM3_CONFIG_DEBUG};
    print "DEBUG: shrdfile='$shrdfile'\n" if $ENV{PPM3_CONFIG_DEBUG};

    # Pick the least public place to save changes.
    my $saveto = $treefile ? $treefile :
		 $userfile ? $userfile : $shrdfile;
    $conf->setfile($saveto);
    $conf->setsave unless $mode eq 'ro';

    # Load the "most private" file.
    return $conf->loadfile($treefile) if -f $treefile && -s _;
    return $conf->loadfile($userfile) if -f $userfile && -s _;
    return $conf->loadfile($shrdfile) if -f $shrdfile && -s _;

    # Neither the shared nor the user's file exists. Let's attempt to
    # create a stub copy of the file, initialised to reasonable defaults.
    print "DEBUG: Writing a stub config file for '$name'.\n"
	if $ENV{PPM3_CONFIG_DEBUG};
    eval {
	# Create config dir _even_ if we're going to load the file ro.
	my $stubfile = $conf->file;
	local *FILE;
	open (FILE, "> $stubfile") or die $!;	# caught by the eval
	print FILE config_file_stub($orig);	# write stub config
	close FILE or die $!;
	$conf->loadfile($stubfile);
    };
    if ($@) {
	die "Fatal error: couldn't find or create config file $name: $@";
    }

    return $conf;
}

# Returns the "tree" configuration directory. This is the directory used by
# 'ppminst'.
sub tree_conf_dir {
    my $d = $ENV{PPM3_PERL_SITELIB}
            || do { require Config; $Config::Config{sitelibexp} };
    return "$d/ppm-conf";
}

sub get_tree_conf_dir {
    return mymkpath(tree_conf_dir());
}

# Returns the user's configuration directory. Note: throws an exception if the
# directory doesn't exist and cannot be created.
sub get_user_conf_dir {
    return undef unless $ENV{PPM3_USER};
    return mymkpath(join PATHSEP, licGetHomeDir(), CONFDIR);
}

# Returns the shared configuration directory. Note: throws no exception, but
# the directory is not guaranteed to exist. Install scripts and such should be
# sure to create this directory themselves.
sub get_shared_conf_dir {
    return undef unless $ENV{PPM3_SHARED};
    return join PATHSEP, UNIX_SHARED_ROOT, KEYDIR, CONFDIR
      if !WIN32;

    my ($R,%R);
    require Win32::TieRegistry;
    Win32::TieRegistry->import(TiedHash => \%R);
    bless do { $R = \%R }, "Win32::TieRegistry";
    $R->Delimiter('/');
    my $wkey = $R->{"HKEY_LOCAL_MACHINE/SOFTWARE/Microsoft/Windows/"};
    my $xkey = $wkey->{"CurrentVersion/Explorer/Shell Folders/"};
    my $shared_root = $xkey->{"/Common AppData"};
    return join PATHSEP, $shared_root, KEYDIR, CONFDIR;
}

sub get_conf_dirs {
    my @path;
    push @path, get_shared_conf_dir(), get_user_conf_dir();
    @path
}

#=============================================================================
# licGetHomeDir(): copied and converted from the Licence_V8 code:
#=============================================================================
sub licGetHomeDir {
    my $dir;
    my ($env1, $env2);

    if (WIN32) {
	require Win32;
	if (defined &Win32::GetFolderPath) {
	    $env1 = Win32::GetFolderPath(Win32::CSIDL_APPDATA());
	}
	$env1 = $ENV{APPDATA} unless defined $env1;
    }

    unless ($env1) {
	$env1 = $ENV{HOME};
    }

    # On Linux & Solaris:
    if (!WIN32) {
	unless ($env1) {
	    $env1 = (getpwuid $<)[7]; # Try to get $ENV{HOME} the hard way
	}
	$dir = sprintf("%s/.%s", $env1, KEYDIR);
    }

    # On Windows:
    else {
	unless ($env1) {
	    $env1 = $ENV{USERPROFILE};
	}
	unless ($env1) {
	    $env1 = $ENV{HOMEDRIVE};
	    $env2 = $ENV{HOMEPATH};
	}
	unless ($env1) {
	    $env1 = $ENV{windir};
	}
	unless ($env1) {
	    die ("Couldn't find HOME / USERPROFILE / HOMEDRIVE&HOMEPATH / windir");
	}
	$env2 ||= "";
	$dir = $env1 . $env2;
	$dir =~ s|/|\\|g;

	# Win32 _stat() doesn't like trailing backslashes, except for x:\
	while (length($dir) > 3 && substr($dir, -1) eq '\\') {
	    chop($dir);
	}

	die ("Not a directory: $dir") unless -d $dir;

	$dir .= PATHSEP;
	$dir .= KEYDIR;
    }

    # Create it if it doesn't exist yet
    return mymkpath($dir);
}

sub repository {
    my $rep  = shift;

    #No ppm repositories for these platforms
    return if SOLARIS_X86 or AIX;

    my $ver = $^V ? sprintf("%vd", substr($^V,0,2)) : $];

    my $ppm3 = "http://ppm.ActiveState.com/PPM/ppmserver%s.plex?urn:/PPM/Server/SQL";
    my $www  = "http://ppm.ActiveState.com/PPMPackages/%s";

    my $verplat1 = "";
    my $verplat2 = $ver;

    if ($^V and $^V ge v5.8.0) {
	my %osmap = (MSWin32 => "windows");
	my $plat = $osmap{$^O} || $^O;
	$verplat1 = "-$ver-$plat";
	$verplat2 = "$ver-$plat";
    }
    my %reps = (
	ppm3 => sprintf($ppm3, $verplat1),
	www  => sprintf($www, $verplat2),
    );
    return $reps{$rep};
}

sub config_file_stub {
    my $name = shift;
    if ($name eq 'clientlib') {
	my $tmp = $ENV{TEMP} || WIN32 ? 'C:\TEMP' : '/tmp';
	my $server = repository('ppm3');
	return <<END;
downloadbytes: 16384
profile_enable: 0
profile_server: $server
rebuildhtml: 0
tempdir: $tmp
tracefile: ppm3.log
tracelvl: 0
END
    }
    elsif ($name eq 'cmdline') {
	# This is actually a little bit wrong, since there are (potentially)
	# multiple frontends. Each frontend should really be responsible for
	# its own configuration data. Still, I don't care all that much.
	return <<'END';
case-sensitivity: 0
fields: name version abstract
follow-install: 1
force-install: 0
install-verbose: 1
max_history: 100
page-lines: 24
pager: ""
prompt-context: 0
prompt-slotsize: 11
prompt-verbose: 0
remove-verbose: 1
sort-field: name
upgrade-verbose: 1
verbose-startup: 1
END
    }
    elsif ($name eq 'instkey') {
	my $txt = do {
	    require PPM::Sysinfo;
	    my $DATA = PPM::Sysinfo::generate_inst_key();
	    return PPM::YAML::serialize($DATA);
	};
    }
    elsif ($name eq 'repositories') {
	my $repositories;
	if (my $url_ppm3 = repository('ppm3')) {
		$repositories .= <<END;
ActiveState Package Repository: %
    url: $url_ppm3
END
	}
    
    return $repositories;
    }
    elsif ($name eq 'targets') {
	# Targets.cfg is the only oddball, because there are multiple targets
	# out there. What we do is this: at build time (at ActiveState) we
	# write a targets.cfg file which is to be relocated at install time.
	# Instead of installing it using a post-install script, ppm3-bin will
	# now look for it next to the binary, and use its contents as the stub
	# targets.cfg file. If the user deletes their .ActiveState/PPM
	# directory, this will magically reappear.
	require FindBin;
	my $f = "$FindBin::Bin/ppm3-bin.cfg";
	my $txt;
	if (-f $f) {
	    local *STUB;
	    open STUB, $f	or die "can't open $f: $!";
	    $txt = do { local $/; <STUB> };
	    close STUB		or die "can't close $f: $!";
	}
	return $txt;
    }
    return '';	# unrecognized file
}

1;
